homeassistant/helpers/event.py

```diff
@@ -88,21 +93,24 @@ def async_track_state_change(
     @callback
     def state_change_listener(event: Event) -> None:
         """Handle specific state changes."""
-        old_state = event.data.get("old_state")
-        if old_state is not None:
-            old_state = old_state.state
-
-        new_state = event.data.get("new_state")
-        if new_state is not None:
-            new_state = new_state.state
-
-        if match_from_state(old_state) and match_to_state(new_state):
```

Comment:
In the old case, we would run both matchers. If either `from_state` or `to_state` is `None`, the matcher would be a lambda that returns `True`. However, we would still run the matcher on the other state value.

In this new logic it looks like we only run matchers if both are not None?

Reply:
Looks like I need to add more tests :)

Reply:
Fixed and tests added